lib/Optimizer/CodeGen/LowerToQIR.cpp

```diff
@@ -503,13 +503,6 @@ class OneTargetOneParamRewrite : public ConvertOpToLLVMPattern<OP> {
     auto instName = instOp->getName().stripDialect().str();
     auto numControls = instOp.getControls().size();
 
-    // TODO: handle general control ops. For now, only allow Rn with 1
```

Comment:
The rest of this implementation is specific to a single control (see line 546 there `auto control = *instOp.getControls().begin();`, it checks for either a Veq or a Ref). It will need to be updated to handle things like `ry<ctrl>(angle, q0, q1, q2, qTarget)` or `ry<ctrl>(angle, veq1, veq2, q0, q1, qTarget)`

Reply:
Ok I've been very skeptical of this as well, but believe it or not, those signatures actually work...

This is from a new test in the C++ builder tester
```
    auto [kernel, angle] = cudaq::make_kernel<float>();
    auto controls1 = kernel.qalloc(3);
    auto controls2 = kernel.qalloc(3);
    auto control3 = kernel.qalloc();
    auto control4 = kernel.qalloc();
    auto control5 = kernel.qalloc();
    auto target = kernel.qalloc();

    // Place the target in the 1-state.
    kernel.x(target);

    // rx
    kernel.rx<cudaq::ctrl>(M_PI, controls1, controls2, control3, control4,
                           control5, target);
    kernel.rx<cudaq::ctrl>(angle, control5, control4, control3, controls2,
                           controls1, **target);
```

which produces the following quake
```
 quake.rx (%cst) [%0, %1, %2, %3, %4] %5 : (f64, !quake.veq<3>, !quake.veq<3>, !quake.ref, !quake.ref, !quake.ref, !quake.ref) -> ()
 quake.rx (%arg0) [%4, %3, %2, %1, %0] %5 : (f32, !quake.ref, !quake.ref, !quake.ref, !quake.veq<3>, !quake.veq<3>, !quake.ref) -> ()
```

Reply:
I would expect the quake to be constructed correctly. But then what happens when you JIT compile / execute the kernel? Did that work? Reading that LowerToQIR pass I would think that the LLVM produced is not valid.

Reply:
I thought the same, but I can't find any discrepancies in the simulated output. Check out the new `BuilderTester.checkTemporary` that I added, everything lines up the way I'd expect

Reply:
Would you mind printing the LLVM that this quake produces and pasting it here?

Reply:
Ok I think I found the issue, and it lines up with what you're pointing to. We're producing the correct Quake, but it's only using the front qubit as a control when simulating. See

```
  {
    auto kernel = cudaq::make_kernel();
    auto controls1 = kernel.qalloc(2);
    auto controls2 = kernel.qalloc(2);
    auto control3 = kernel.qalloc();
    auto target = kernel.qalloc();

    // Every control but the first in 1-state
    kernel.x(controls2);
    kernel.x(control3);

    kernel.rx<cudaq::ctrl>(M_PI, controls1, controls2, control3, target);

    auto counts = cudaq::sample(kernel);
    counts.dump();

    // Target qubit should've stayed in |0>
    EXPECT_EQ(counts.count("001110"), 1000);
  }
```

But when we rotate only the first controls (`kernel.x(controls1);` only), we get the unexpected output state of "110001"

Reply:
Thanks for pointing this out, my "fix" felt wrong but I couldn't track down any issues with it. Fixing it now 😄

Reply:
Not to cause any brain damage, but quake will allow controls of type ref, veq, control, and wire. In lowering to QIR, we probably should deal with the first two types, but reject the last two with an error.

Reply:
@schweitzpgi Silly question, but is there a simple way to return the type of "Type" as a string? It'd be nice to merge these

```
    if (type.isa<quake::ControlType>()) {
      return instOp.emitError("unsupported control type for " + instName +
                              ": `quake::ControlType`");
    }
    if (type.isa<quake::WireType>()) {
      return instOp.emitError("unsupported control type for " + instName +
                              ": `quake::WireType`");  
    }
```

into one `(!type.isa<quake::VeqType>() || !type.isa<quake::RefType>())` instead of hard-coding the error messages

Reply:
```c++
isa<quake::VeqType, quake::RefType>(type)
```
should even work.

Reply:
@amccaskey Circling back here, this code actually doesn’t execute on QPP for our standard gates, e.g X (I haven’t tested on custatevec yet)
```
  {
    auto kernel = cudaq::make_kernel();
    auto controls1 = kernel.qalloc(3);
    auto controls2 = kernel.qalloc(3);
    auto control3 = kernel.qalloc();
    auto target = kernel.qalloc();
    
    // Returns qpp dim error
    kernel.x<cudaq::ctrl>(controls1, controls2, control3, target);

    // Doesn't return any error but has the incorrect output
    kernel.rx<cudaq::ctrl>(3.1415, controls1, controls2, control3, target);

    auto counts = cudaq::sample(kernel);
    counts.dump();
  }
  ```

Running into the infamous
```C++ exception with description "qpp::applyCTRL(): Subsystems mismatch dimensions! [ctrl/dims]" thrown in the test body.```

A more simple example like `kernel.x<cudaq::ctrl>(controls1, target);` runs fine. 

Interestingly, the rx gate shows the unwanted behavior we discussed above, but without returning an error. Not sure if I’m missing something here?

Reply:
@amccaskey 
Ok all of this has been addressed, and the more advanced function signatures like
```
    kernel.rx<cudaq::ctrl>(M_PI, controls1, controls2, control3, control4,
                           control5, target);
```
have their own tests in the QuantinuumTester. I believe there is a more general bug in library mode when using a signature like that with any control gate. Will be filing an issue shortly.